static/monitoring.js

```diff
@@ -109,6 +112,65 @@ function appendLine(element, text) {
     element.appendChild(document.createElement("br"));
 }
 
+
+function fetchHistory(parent, directive, page=0) {
+    parent.childNodes.forEach(child => {
+        child.remove();
+    });
+    const parentdata = parent.dataset;
+    if (page > 0) {
+        parent.appendChild(create("button", "Last 20", "page_history_prev"));
+    }
+    fetch("/api/attestation_history.json", {method: "POST", body: {
+        token: parentdata.token,
+        fingerprint: parentdata.fingerprint,
+        offsetId: parentdata.offsetId,
+        directive
+    },
+        credentials: "same-origin"}).then(response => {
+        if (!response.ok) {
+            return Promise.reject();
+        }
+        return response.json();
+    }).then(attestations => {
+        for (const attestation of attestations) {
+            history.appendChild(create("h4", new Date(attestation.time)));
+
+            const p = history.appendChild(document.createElement("p"));
+            const result = attestation.strong ?
+                "Successfully performed strong paired verification and identity confirmation." :
+                "Successfully performed basic initial verification and pairing.";
+            p.appendChild(create("strong", result));
+
+            history.appendChild(create("h5", "Verified device information (constants omitted):"));
+            history.appendChild(create("p", attestation.teeEnforced));
+            history.appendChild(create("h5", "Information provided by the verified OS:"));
+            history.appendChild(create("p", attestation.osEnforced));
+        }
+        function fetchHistoryNextPage() {
+            return fetchHistory(parent, ATTESTATION_HISTORY_PAGINATE_NEXT, parentdata.page + 1);
+        }
+
+        function fetchHistoryPrevPage() {
+            return fetchHistory(parent, ATTESTATION_HISTORY_PAGINATE_BACK, parentdata.page - 1);
+        }
+        if (attestations.length === 20) {
+            parent.appendChild(create("button", "Next 20", "page_history_next"));
+            parentdata.offsetId = attestations[attestations.length - 1].offsetId;
+        }
+        parentdata.offsetId = attestations[attestations.length - 1].offsetId;
+        const next_page = document.getElementsByClassName("page_history_next")[0];
```

Comment:
I don't see how this can work. You're fetching all elements with the class `page_history_next`, not just within a specific scope.

Reply:
There should only be one such element though since the child nodes in the history DOM gets cleared out.